econml/shap.py

```diff
@@ -0,0 +1,263 @@
+    models: a single estimator or a list of estimators with one estimator per treatment
+        models for the model's final stage model.
+    X: (m, d_x) matrix
+        Features for each sample. Should be in the same shape of fitted X in final stage.
+    d_t: tuple of int
+        Tuple of number of treatment (exclude control in discrete treatment scenario).        
+    d_y: tuple of int
+        Tuple of number of outcome.
+    feature_names: optional None or list of strings of length X.shape[1] (Default=None)
+        The names of input features.
+    treatment_names: optional None or list (Default=None)
+        The name of treatment. In discrete treatment scenario, the name should not include control name.
+    output_names:  optional None or list (Default=None)
+        The name of the outcome.
+
+    Returns
+    -------
+    shap_outs: nested dictionary of Explanation object
+        A nested dictionary by using each Y and each T as key and the shap_values explanation object as value.
+    """
+
+    (dt, dy, treatment_names, output_names) = _define_names(d_t, d_y, treatment_names, output_names)
+    if not isinstance(models, list):
+        models = [models]
+    assert len(models) == dt, "Number of final stage models don't equals to number of treatments!"
+    # define masker by using entire dataset, otherwise Explainer will only sample 100 obs by default.
+    background = shap.maskers.Independent(X, max_samples=X.shape[0])
+
+    shap_outs = defaultdict(dict)
+    for i in range(dt):
+        try:
+            explainer = shap.Explainer(models[i], background,
+                                       feature_names=feature_names)
+        except Exception as e:
+            print("Final model can't be parsed, explain const_marginal_effect() instead!")
+            return _shap_explain_cme(cme_model, X, d_t, d_y, feature_names, treatment_names,
+                                     output_names)
+        shap_out = explainer(X)
+        if dy > 1:
+            for j in range(dy):
+                base_values = shap_out.base_values[..., j]
+                values = shap_out.values[..., j]
+                main_effects = None if shap_out.main_effects is None else shap_out.main_effects[..., j]
+                shap_out_new = Explanation(values, base_values=base_values,
+                                           data=shap_out.data, main_effects=main_effects,
+                                           feature_names=shap_out.feature_names)
+                shap_outs[output_names[j]][treatment_names[i]] = shap_out_new
+        else:
+            shap_outs[output_names[0]][treatment_names[i]] = shap_out
+
+    return shap_outs
+
+
+def _shap_explain_dml_model_cate(model_final, X, T, d_t, d_y, fit_cate_intercept,
+                                 feature_names=None, treatment_names=None, output_names=None):
+    """ 
+    The method to explain `model_cate` of parametric final stage DML using shap Explainer()
```

Comment:
Maybe the docstring could explain what types of cate models this explains, ie.. linear models that were fitted on the cross_product of featurizer(X) and T.

This way if we ever need this functionality somwhere else we can still re-use it (e.g. in LinearDMLIV etc.)

Reply:
Done